### PR TITLE
Adding ISAAC Plugin

### PR DESCRIPTION
**Pull request by theZiz**
This is a pull request to add support for the In-Situ visualization library ISAAC to PIConGPU.

I tried to adress all issues mentioned here https://github.com/theZiz/picongpu/commit/70c0593ac00a58db88fca3b6fcf72aeadf1660f6 .

### CODE REVIEW

File: src/picongpu/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +122,37 @@ if( (CUDA_VERSION VERSION_GREATER 7.0) AND
 endif()
 
 ################################################################################
+# ISAAC
+################################################################################
+
+find_package(ISAAC 0.1.0 CONFIG)
+if(ISAAC_FOUND)
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: `AND CMAKE_CXX_STANDARD NOT EQUAL 98`

Comment (theZiz) [inline src/picongpu/CMakeLists.txt]: Done. https://github.com/ComputationalRadiationPhysics/picongpu/pull/1474/commits/badeb63715fc4cc949b61fa29d364298693e5b87

**Comment by ax3l** [comment]
I can lead the review, @psychocoderHPC please have a sweep how you would like to see the new pluging in the C++ backend.

- [ ] [ISAAC wiki article](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Plugin%3A-ISAAC) is coming, yeah!

- [ ] We should also add notes to the `INSTALL.md` and add isaac to the compile suite (C++11, after next release).

File: src/picongpu/CMakeLists.txt [hunk 2]
```diff
@@ -122,6 +122,44 @@ if( (CUDA_VERSION VERSION_GREATER 7.0) AND
...
+        set(LIBS ${LIBS} ${ISAAC_LIBRARIES})
+
+        set(ISAAC_MAX_FUNCTORS "3" CACHE STRING "Max length of the isaac functor chain" )
+        set(ISAAC_DEFAULT_WEIGHT "7" CACHE STRING "Default weight of an isaac source" )
+
+        add_definitions(${ISAAC_DEFINITIONS})
+        add_definitions(-DISAAC_MAX_FUNCTORS=${ISAAC_MAX_FUNCTORS})
+        add_definitions(-DISAAC_FUNCTOR_POW_ENABLED=0)
+        add_definitions(-DISAAC_DEFAULT_WEIGHT=${ISAAC_DEFAULT_WEIGHT})
+
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: pls remove one of those two newlines :)

File: src/picongpu/CMakeLists.txt [hunk 3]
```diff
@@ -244,6 +282,10 @@ else()
     add_definitions(-DBOOST_RESULT_OF_USE_TR1)
 endif()
 
+if( CMAKE_CXX_STANDARD EQUAL 11 )
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: why is this needed? (long explanation here + inline comment)

File: src/picongpu/include/plugins/IsaacPlugin.hpp [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+        void init(MappingDesc *cellDescription, bool movingWindow)
+        {
+            this->cellDescription = cellDescription;
+            this->movingWindow = movingWindow;
+        }
+
+        static std::string getName()
+        {
+            return ParticleType::FrameType::getName() + std::string(" density");
```
Comment (ax3l) [inline src/picongpu/include/plugins/IsaacPlugin.hpp]: `FieldTmpOperation< FrameSolver, ParticleType >` should also have a `getName`() you can use.

File: src/picongpu/CMakeLists.txt (hunk 3 above)
Comment (theZiz) [inline src/picongpu/CMakeLists.txt]: http://stackoverflow.com/questions/15634114/cant-link-program-using-boost-filesystem

File: src/picongpu/include/plugins/IsaacPlugin.hpp (hunk 4 above)
Comment (theZiz) [inline src/picongpu/include/plugins/IsaacPlugin.hpp]: Nope:
```…/picongpu/include/plugins/IsaacPlugin.hpp(128): error: class "picongpu::FieldTmpOperation<picongpu::particleToGrid::ComputeGridValuePerFrame<picongpu::UsedParticleShape, picongpu::particleToGrid::derivedAttributes::ChargeDensity>, picongpu::PIC_Electrons>" has no member "getName"```

Comment (ax3l) [inline src/picongpu/include/plugins/IsaacPlugin.hpp]: `FrameSolver` has it. See [`hdf5::class WriteFields<FieldTmpOperation<Solver, Species> >`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/9612cede86fa0c93f531d80465ea95d1d531c7ce/src/picongpu/include/plugins/hdf5/WriteFields.hpp#L148-L155)

Comment (theZiz) [inline src/picongpu/include/plugins/IsaacPlugin.hpp]: Fixed btw.

File: src/picongpu/CMakeLists.txt (hunk 3 above)
Comment (theZiz) [inline src/picongpu/CMakeLists.txt]: I did just remove this line as this problem will not occur anymore if the minimal requirement of PIConGPU is updated to Boost 1.57.

**Comment by ax3l** [comment]
@theZiz can you please address the two points mentioned above in a follow-up and finish the wiki if anything is missing?

https://github.com/ComputationalRadiationPhysics/picongpu/pull/1474#issuecomment-221037024

**Comment by theZiz** [comment]
Yes. Tomorrow ;)

**Comment by ax3l** [comment]
also, the CMake output on "not found" for ISAAC is too verbose.

We do only write a `STATUS` on missing optional dependencies/plugins, not a warning as it is right now:
```
CMake Warning at src/picongpu/CMakeLists.txt:124 (find_package):
  Could not find a package configuration file provided by "ISAAC" (requested
  version 0.1.0) with any of the following names:

    ISAACConfig.cmake
    isaac-config.cmake

  Add the installation prefix of "ISAAC" to CMAKE_PREFIX_PATH or set
  "ISAAC_DIR" to a directory containing one of the above files.  If "ISAAC"
  provides a separate development package or SDK, be sure it has been
  installed.
```

**Comment by ax3l** [comment]
The short-cut that one can actually install most dependencies via the provided git submodules seems also to be missing in both the ISAAC `INSTALL.md` and the PIConGPU plugin wiki (and I just realized this again after installing all of them by hand...)

**Comment by ax3l** [comment]
the ISAAC section is also surprisingly high in the main `CMakeLists.txt` file for a plugin.

**Comment by theZiz** [comment]
About the high of ISAAC: Historic reason. Whereby someone could ask, why Vampir is so high, too. Probably because it is no real plugin... Anyway: Shall I lower ISAAC? Is this a real issue?

**Comment by theZiz** [comment]
I created a new pull request for the requested changes. If you find more bugs / nonsense, please discuss it there, so I can fix it for the next merge: https://github.com/ComputationalRadiationPhysics/picongpu/pull/1630

**Comment by theZiz** [comment]
Regarding https://github.com/ComputationalRadiationPhysics/picongpu/pull/1474#issuecomment-221037024 : I hopefully fixed the mentioned issues.

**Comment by ax3l** [comment]
ty! :)
